gh-107421: Clarify `OrderedDict` Examples and Recipes

shailshouryya opened the pull request:
<strike>
In addition to updating the documentation to include the changes suggested in https://github.com/python/cpython/issues/107421, this PR also makes a minor fix in commit ff22bac665c0520fc27d16bc254563adf30afc06 to ensure the <code>MultiHitLRUCache</code> never has more than <code>maxsize</code> items in the <code>cache</code> attribute at any given time (see the commit message for more information).
</strike>
<br>

See https://github.com/python/cpython/pull/107613#discussion_r1284124385 for an explanation about why "never [having] more than `maxsize` items in the `cache` attribute at any given time" might not be a good idea.

<!-- gh-issue-number: gh-107421 -->
* Issue: gh-107421
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--107613.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

Here's the direct link to the preview of the [`OrderedDict` Examples and Recipes section](https://cpython-previews--107613.org.readthedocs.build/en/107613/library/collections.html#ordereddict-examples-and-recipes)

shailshouryya committed:
Use explicit `last=False` argument for `.popitem` method

shailshouryya committed:
Remove unnecessary default argument for `self.requests.pop` call in `MultiHitLRUCache`

See explanation at https://github.com/python/cpython/issues/107421#issue-1827368825
for more information.

shailshouryya committed:
Add comments to `__call__` method `MultiHitLRUCache` for clarity

shailshouryya committed:
Maintain invariant of `maxsize` items in `cache` attribute in `MultiHitLRUCache`

This minor change ensures the cache never has more than `maxsize`
items in `cache`. The previous implementation would temporarily
exceed `maxsize` items by one item when the cache adds the
newest entry to `cache` before removing the least recently used
item from `cache` with `self.cache.popitem(last=False)`.

rhettinger commented on Doc/library/collections.rst:
```diff
@@ -1252,16 +1252,17 @@ variants of :func:`functools.lru_cache`:
                 return self.cache[args]
             result = self.func(*args)
-            self.requests[args] = self.requests.get(args, 0) + 1
+            self.requests[args] = self.requests.get(args, 0) + 1 # increment existing value by 1 if seen before, otherwise set to 1
```
This new comment does not add value.  Please leave the line as-is.

shailshouryya replied:
I added this comment when I was a little too focused on issue 107421, but looking at this code block again, I agree that this comment is pretty much just restating the obvious.

I updated this part of the example in commit ecbb79a2ae723a7e8baf57bce1f3a05feaaab4db.

rhettinger commented on Doc/library/collections.rst:
```diff
@@ -1252,16 +1252,17 @@ variants of :func:`functools.lru_cache`:
+                    self.requests.popitem(last=False)
             else:
-                self.requests.pop(args, None)
+                # entry has been seen more than cache_after times
```
This is comment is clear; however, I'm dubious that it adds value because the code already reads as "if requests[args] < cache_after: ... else: ..."

There isn't much of a point in having a comment that says exactly what the code already says.

shailshouryya replied:
I added this comment using the same [reasoning I mentioned above](https://github.com/python/cpython/pull/107613#discussion_r1284705872), and agree this comment is also repetitive.

I also updated this part of the example in commit ecbb79a2ae723a7e8baf57bce1f3a05feaaab4db.

rhettinger commented on Doc/library/collections.rst:
```diff
@@ -1252,16 +1252,17 @@ variants of :func:`functools.lru_cache`:
             else:
-                self.requests.pop(args, None)
+                # entry has been seen more than cache_after times
+                self.requests.pop(args) # no longer need to keep track of how many times this entry has been seen
```
This comment is reasonable and add some value.  Please move it to a separate line, just before the relevant code. 

Also, keep the original `None` argument.  It provides some safety in a multithreaded environment.  We never want this line to raise an exception.

shailshouryya replied:
I moved this comment to the lines just above the code in commit 77f10b2a66b2d7353ccb13783021b459484e0de4.

I also added context the multi-threading problems that could arise that you mentioned in commit a03483d5cd9cc3956f4f4bae51c416dc67d1fb96, since I think the race conditions that could occur in multi-threading environments is kind of obvious for someone who is already designing code for multi-threading environments - but might be missed by someone who _**initially designs the code for a single-threaded environment**_ and then moves it into a multi-threaded environment without reviewing all the code.

I think this is useful context and might help some people avoid problems if they mistakenly take out the default argument from `pop(args, None)` because they think it is unnecessary (in their single-threaded environment), but let me know if you think the comment is too long/should be changed, or if the comment is unnecessary altogether.

rhettinger replied:
Please remove the comments altogether.   Just stick with the 1/0 to True/False edit.

shailshouryya replied:
I think this is related to https://github.com/python/cpython/pull/107613#discussion_r1284981437, in which case, commit aae218b39b27f09a39ed0cd6fba09cee26aafbea should take care of this as well.

rhettinger commented on Doc/library/collections.rst:
```diff
@@ -1252,16 +1252,17 @@ variants of :func:`functools.lru_cache`:
-                self.requests.pop(args, None)
+                # entry has been seen more than cache_after times
+                self.requests.pop(args) # no longer need to keep track of how many times this entry has been seen
+                if len(self.cache) == self.maxsize:
```
I do not agree with this code edit.  Letting the cache grow larger than the maxsize and then popping is safer because we are guaranteed that at least one entry is present. Also, we tend to put deletion code after the insertion code so that the cache is in a consistent state prior to deletion whose DECREF logic can trigger arbitrary Python code.  

Please restore the original code order.

shailshouryya replied:
I reverted this change in commit 26f9f9061d4b196845fd82da3738ccddcfa7ab7c.

I think I got the gist of what you're saying here, but I don't think I understand the internals of CPython enough to follow all of it.

> Letting the cache grow larger than the maxsize and then popping is safer because we are guaranteed that at least one entry is present.

Is this also to avoid issues in multi-threaded environments, as you mentioned above (and in issue 107421)?

> Also, we tend to put deletion code after the insertion code so that the cache is in a consistent state prior to deletion whose DECREF logic can trigger arbitrary Python code.

I think this is referring to garbage collection that might remove the cache when the cache becomes empty, correct? In that case, wouldn't the `self.cache` still be safe from garbage collection since the reference count is still (at least) 1 since the instance has a reference to the `self.cache` attribute? Or is this implementation dependent? (Or is it something else altogether? 😅)

shailshouryya committed:
Remove comments that restate what the code does

These comments just say what the code is doing in different
words and do not add any new information or context.

shailshouryya committed:
Move useful comment to separate lines before the relevant code

shailshouryya committed:
Keep `pop(args, None)` for simplicity in multi-threaded environments

See the comment at
https://github.com/python/cpython/pull/107613/files#r1284119710 and
https://github.com/python/cpython/issues/107421#issuecomment-1665239309
for more information.

shailshouryya committed:
Revert "Maintain invariant of `maxsize` items in `cache` attribute in `MultiHitLRUCache`"

This reverts commit ff22bac665c0520fc27d16bc254563adf30afc06.
See the explanation at
https://github.com/python/cpython/pull/107613/files#r1284124385
for more information.

shailshouryya committed:
Remove backticks from code comment to avoid linter error

See https://github.com/python/cpython/actions/runs/5765119619/job/15630465684?pr=107613.

shailshouryya committed:
Remove backticks from other code comment to avoid linter error

See https://github.com/python/cpython/actions/runs/5765303505/job/15631004596?pr=107613.
This was missed and should have been included in the previous commit.

rhettinger commented on Doc/library/collections.rst:
```diff
@@ -1256,12 +1256,18 @@ variants of :func:`functools.lru_cache`:
+                    self.requests.popitem(last=False)
             else:
+                # no longer need to keep track of how many times this
+                # entry has been seen; in a single-threaded environment,
```
Please remove 1261 to 1265. The comments are getting in the way of reading the code.  Please stick to the substance of the original issue.

shailshouryya replied:
Updated in commit aae218b39b27f09a39ed0cd6fba09cee26aafbea

shailshouryya replied:
I'm reading through https://discuss.python.org/t/adopting-the-diataxis-framework-for-python-documentation/15072, and I'm wondering if we could add something along the lines of these comments elsewhere in the documentation (with a new corresponding issue).

Do you think it is worth it to include information about being careful about race conditions in multi-threaded environments, either at the top of this page (The Python Standard Library » Data Types» collections — Container datatypes) or in some other documentation page?

E.g. something like, "In a single-threaded environment, using the data structures provided by the `collections` module is straightforward, but in a multi-threaded environment, take care to think about how race conditions may cause unwanted behavior in an insertion, modification, or deletion operation for the data structure you are using."

rhettinger replied:
This is a really minor nuance.  Let's let it be.

shailshouryya committed:
Remove all comments

See https://github.com/python/cpython/pull/107613/files#r1284981437.
